Add a ModifyNodeGenerator

socratesgorilla opened the pull request:
Adds a new mesh generator with the ability to modify existing nodes. Closes #18577

Commit:
Added ModifyNodeGenerator refs #18577

Delete testCart.i

Update tests

Fixed whitespace

clang reformat

Update tests

Update ModifyNodeGenerator.C

Update framework/include/meshgenerators/ModifyNodeGenerator.h

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

Update framework/src/meshgenerators/ModifyNodeGenerator.C

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

Update ModifyNodeGenerator.C

reformat

Fixed import issues

Added gold file for test

Update framework/src/meshgenerators/ModifyNodeGenerator.C

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

Update framework/include/meshgenerators/ModifyNodeGenerator.h

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Yaqi edits and comment changes

Fixed test_in exodiff, hopefully

Hopefully fix the last test issue

Debugging distributed mesh errors
Make mesh new positions more visible in the diffusion test for the modified mesh
Refs #18577

rename everything

Added mesh_type

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/795852/) on 86a37c0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18590/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18590/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0e5f0880f25b7c7512874866fb9a1e9c63f2eea2`

dschwen commented on framework/src/meshgenerators/ModifyNodeGenerator.C:
```diff
@@ -0,0 +1,54 @@
+
+  params.addRequiredParam<std::vector<Point>>("new_position", "New position in vector space");
+
+  params.addClassDescription("Modifies the position of one node");
```
Obviously modifies the position of not just one node :-)

socratesgorilla replied:
Changed

loganharbour commented on framework/src/meshgenerators/ModifyNodeGenerator.C:
```diff
@@ -0,0 +1,54 @@
+std::unique_ptr<MeshBase>
+ModifyNodeGenerator::generate()
+{
+  if (_node_id.size() != _new_position.size())
+    mooseError("Node ids and new positions sizes do not match. Please check your input");
```
Move this up into the object constructor (checking earlier is better). In addition, use a `paramError` so that it provides some context:

```
if (_node_id.size() != _new_position.size())
  paramError("node_id", "Does not match the length of 'new_position'");
```

loganharbour commented on framework/src/meshgenerators/ModifyNodeGenerator.C:
```diff
@@ -0,0 +1,54 @@
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  // get the node
+  for (int i = 0; i < int(_node_id.size()); i++)
```
```suggestion
  for (MooseIndex(_node_id.size()) i = 0; i < _node_id.size(); i++)
```

`MooseIndex` gets the correct type for iteration

loganharbour commented on framework/include/meshgenerators/ModifyNodeGenerator.h:
```diff
@@ -0,0 +1,37 @@
+  ModifyNodeGenerator(const InputParameters & parameters);
+
+  std::unique_ptr<MeshBase> generate() override;
+
+  Elem * getElemType(const std::string & type);
```
```suggestion
```

YaqiWang commented on framework/src/meshgenerators/ModifyNodeGenerator.C:
```diff
@@ -0,0 +1,54 @@
+
+  params.addRequiredParam<MeshGeneratorName>("input", "Input mesh to modify");
+
+  params.addRequiredParam<std::vector<int>>("node_id", "Id of modified node");
```
Suggest changing `int` to `dof_id_type` to be consistent with its use in `Mesh::query_node_ptr`.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/802511/) on 049480c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18590/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18590/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format eecb6c9188387d801c146253e42b62ed17f9de0f`

loganharbour commented on framework/src/meshgenerators/ModifyNodeGenerator.C:
```diff
@@ -0,0 +1,56 @@
+  // get the node
+  for (MooseIndex(_node_id.size()) i = 0; i < _node_id.size(); i++)
+  {
+    auto * node = mesh->query_node_ptr(_node_id[i]);
+
+    // change the position of the acquired node
+    node->assign(_new_position[i]);
```
```suggestion
    std::size_t num_found = 0;
    auto * node = mesh->query_node_ptr(_node_id[i]);

    // change the position of the acquired node
    if (node)
    {
      node->assign(_new_position[i]);
      ++num_found;
    }
    
    // Make sure we found the node
    mesh->comm().sum(num_found)
    if (!num_found)
      mooseError("A node with the ID ", _node_id[i], " was not found.");
```

One last thing - we do need an error when the node wasn't found. You also need to protect against the node not existing in the case of distributed mesh. The node could exist on a subset of the processors instead of all of them.

loganharbour reviewed (changes requested):
I think with this + @YaqiWang's comment above, this should be good.

GiudGiud commented:
I'm hesitating between ModifyNodeGenerator and MoveNodeGenerator for the name. Any preferences?

GiudGiud commented on framework/include/meshgenerators/ModifyNodeGenerator.h:
```diff
@@ -0,0 +1,35 @@
+  std::unique_ptr<MeshBase> generate() override;
+
+protected:
+  /// Mesh that possibly comes from another generator
```
```suggestion
  /// Mesh coming from another mesh generator
```

GiudGiud commented on framework/include/meshgenerators/ModifyNodeGenerator.h:
```diff
@@ -0,0 +1,35 @@
+  /// Mesh that possibly comes from another generator
+  std::unique_ptr<MeshBase> & _input;
+
+  /// The node id
```
```suggestion
  /// The ids of the nodes to be moved
```

GiudGiud commented on framework/include/meshgenerators/ModifyNodeGenerator.h:
```diff
@@ -0,0 +1,35 @@
+  /// The node id
+  const std::vector<int> _node_id;
+
+  /// The new position of the node
```
```suggestion
  /// The new positions of the node
```

GiudGiud commented on framework/include/meshgenerators/ModifyNodeGenerator.h:
```diff
@@ -0,0 +1,35 @@
+#include "MeshGenerator.h"
+
+/**
+ * Modifies the position of one node
```
```suggestion
 * Modifies the position(s) of one or more nodes
```

GiudGiud commented on framework/src/meshgenerators/ModifyNodeGenerator.C:
```diff
@@ -0,0 +1,56 @@
+ModifyNodeGenerator::ModifyNodeGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _input(getMesh("input")),
+    _node_id(getParam<std::vector<int>>("node_id")),
```
modify here and in the header too for Yaqi s  comment

GiudGiud commented:
One test failure is just a missing reference (=gold) file in the gold/ folder. 
The other is a little weirder. Does it pass on your machine?

dschwen commented:
> I'm hesitating between ModifyNodeGenerator and MoveNodeGenerator for the name. Any preferences?

I'd prefer `MoveNodesGenerator` (or `MoveNodeGenerator`). It's a bit more explicit than "modify" (the properties of a Node are its location and id, and this generator only modifies the position)

YaqiWang commented:
Just remember one thing: after we move the node, do we want to check if the Jacobian of the affected elements are still positive?

GiudGiud commented:
We could but I was going to use this to create bad meshes ^^

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/806456/) on a40780f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18590/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18590/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c8dd43526594c7b520d9c04ac37c2fd0f8a3442c`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/806485/) on f2ec0d8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18590/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18590/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c8dd43526594c7b520d9c04ac37c2fd0f8a3442c`

socratesgorilla committed:
Added ModifyNodeGenerator refs #18577

Update test/tests/meshgenerators/move_node_generator/tests

Update framework/include/meshgenerators/MoveNodeGenerator.h

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update framework/src/meshgenerators/MoveNodeGenerator.C

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

GiudGiud commented on test/tests/meshgenerators/move_node_generator/tests:
```diff
@@ -0,0 +1,27 @@
+      cli_args = '--mesh-only'
+      exodiff = 'test_in.e'
+      recover = false
+      mesh_node = replicated
```
```suggestion
      mesh_mode = replicated
```

moosebuild commented:
Job [Generate and verify coverage](https://civet.inl.gov/job/819316/) on 5aa07da wanted to post the following:


## Framework coverage

No coverage report(s) were found for the base commit b937b54e57.


## Modules coverage

No coverage report(s) were found for the base commit b937b54e57.


## Full coverage reports

<details><summary>Reports</summary>

- framework (FAILED)
- [chemical_reactions](https://mooseframework.inl.gov/docs/PRs/18590/coverage/chemical_reactions)
- [combined](https://mooseframework.inl.gov/docs/PRs/18590/coverage/combined)
- [contact](https://mooseframework.inl.gov/docs/PRs/18590/coverage/contact)
- [external_petsc_solver](https://mooseframework.inl.gov/docs/PRs/18590/coverage/external_petsc_solver)
- [fluid_properties](https://mooseframework.inl.gov/docs/PRs/18590/coverage/fluid_properties)
- [fsi](https://mooseframework.inl.gov/docs/PRs/18590/coverage/fsi)
- [functional_expansion_tools](https://mooseframework.inl.gov/docs/PRs/18590/coverage/functional_expansion_tools)
- [geochemistry](https://mooseframework.inl.gov/docs/PRs/18590/coverage/geochemistry)
- [heat_conduction](https://mooseframework.inl.gov/docs/PRs/18590/coverage/heat_conduction)
- [level_set](https://mooseframework.inl.gov/docs/PRs/18590/coverage/level_set)
- [misc](https://mooseframework.inl.gov/docs/PRs/18590/coverage/misc)
- navier_stokes (FAILED)
- [peridynamics](https://mooseframework.inl.gov/docs/PRs/18590/coverage/peridynamics)
- [phase_field](https://mooseframework.inl.gov/docs/PRs/18590/coverage/phase_field)
- [porous_flow](https://mooseframework.inl.gov/docs/PRs/18590/coverage/porous_flow)
- [ray_tracing](https://mooseframework.inl.gov/docs/PRs/18590/coverage/ray_tracing)
- [rdg](https://mooseframework.inl.gov/docs/PRs/18590/coverage/rdg)
- [richards](https://mooseframework.inl.gov/docs/PRs/18590/coverage/richards)
- [stochastic_tools](https://mooseframework.inl.gov/docs/PRs/18590/coverage/stochastic_tools)
- tensor_mechanics (FAILED)
- [xfem](https://mooseframework.inl.gov/docs/PRs/18590/coverage/xfem)

</details>

This comment will be updated on new commits.

GiudGiud commented on framework/include/meshgenerators/MoveNodeGenerator.h:
```diff
@@ -0,0 +1,35 @@
+#include "MeshGenerator.h"
+
+/**
+ * Modifies the position of one node
```
```suggestion
 * Modifies the position of one or more node(s)
```

moosebuild commented:
Job [External App tests](https://civet.inl.gov/job/819593/) on 6d41a14 : invalidated by @socratesgorilla

moosebuild commented:
Job [End User Experience (Ubuntu)](https://civet.inl.gov/job/819601/) on 6d41a14 : invalidated by @socratesgorilla

moosebuild commented:
Job [End User Experience (CentOS 8)](https://civet.inl.gov/job/819600/) on 6d41a14 : invalidated by @socratesgorilla

loganharbour commented on framework/src/meshgenerators/MoveNodeGenerator.C:
```diff
@@ -0,0 +1,66 @@
+
+    // Make sure we found the node
+    mesh->comm().sum(num_found);
+    if (!num_found)
```
Should really have a test for the failure.

loganharbour commented on test/tests/meshgenerators/move_node_generator/tests:
```diff
@@ -0,0 +1,27 @@
+
+    []
+    [3d]
+      type = 'Exodiff'
+      input = 'test2.i'
+      exodiff = 'test2_out.e'
+
+      detail = 'in a Cartesian 3D mesh.'
```
Why does this test have a solve?

socratesgorilla replied:
@GiudGiud I'm honestly not sure why it was here again, could you expound on this?

GiudGiud replied:
just to check that things still work after the mesh modification. 
I could go either way, but it was important for us to check that the solve works, because it does not if you modify the mesh too bad (inverted elements etc)

moosebuild commented:
Job [Internal App tests](https://civet.inl.gov/job/828972/) on b9ecca2 : invalidated by @GiudGiud

unrelated

socratesgorilla committed:
Added failure tests for incorrect inputs
for move_node_generator

loganharbour commented on framework/src/meshgenerators/MoveNodeGenerator.C:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    // Make sure we found the node
+    mesh->comm().sum(num_found);
```
We probably should have done this comm step at the end.... but this is fine
